src/server/pps/server/worker_rc.go

```diff
@@ -839,9 +857,42 @@ func (kd *kubeDriver) getWorkerOptions(ctx context.Context, pipelineInfo *pps.Pi
 		schedulingSpec:        pipelineInfo.Details.SchedulingSpec,
 		podSpec:               pipelineInfo.Details.PodSpec,
 		podPatch:              pipelineInfo.Details.PodPatch,
+		tolerations:           tolerations,
 	}, nil
 }
 
+// transformToleration transofrms a pps.Toleration into a k8s Toleration.  It's used while creating
+// the RC, and at pipeline submission to validate the provided tolerations before saving the
+// pipeline.  It's in here and not in src/pps/pps.go so that the protobuf library doesn't depend on
+// k8s.
+func transformToleration(in *pps.Toleration) (v1.Toleration, error) {
+	var out v1.Toleration
+	out.Key = in.GetKey()
+	out.Value = in.GetValue()
+	if ts := in.GetTolerationSeconds(); ts != nil {
+		out.TolerationSeconds = &ts.Value
+	}
+	switch in.GetEffect() { //exhaustive:enforce
+	case pps.TaintEffect_ALL_EFFECTS:
+		out.Effect = ""
+	case pps.TaintEffect_NO_EXECUTE:
+		out.Effect = v1.TaintEffectNoExecute
+	case pps.TaintEffect_NO_SCHEDULE:
+		out.Effect = v1.TaintEffectNoSchedule
+	case pps.TaintEffect_PREFER_NO_SCHEDULE:
+		out.Effect = v1.TaintEffectPreferNoSchedule
+	}
+	switch in.GetOperator() { //exhaustive:enforce
+	case pps.TolerationOperator_EMPTY:
+		return out, errors.New("cannot omit key/value comparison operator; specify EQUAL or EXISTS")
+	case pps.TolerationOperator_EXISTS:
```

Comment:
Based on the docs, it says `If the operator is Exists, the value should be empty, otherwise just a regular string.` Should we enforce that here by setting `out.Value = ""` just to be safe?

Reply:
I'd rather let k8s handle that logic.  It's something that could change, and we shouldn't be enforcing it.  In general, where the API allows users to represent garbage, we should pass it through.  (A possible case is an admission controller that uses that field for some undocumented use; might as well not break that even though it's a terrible design on their part.)